### PR TITLE
Refactor keyfctl to improve usability and stability

### PR DESCRIPTION
**Pull request by brownjohnf**
This is almost a rewrite of keyfctl that *drastically* improves its usability and reliability. It also lays the foundation for direct deploy support from keyfctl via adapters. I've skeleton adapters as examples for fleet and k8s, but they don't actually do deploys yet. Presumably the fleet one would consume resinctl as a library to do its heavy lifting. I think I'll add that soon, since it should be pretty easy.

Change-Type: major

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -2,6 +2,18 @@
 
 **WARNING: This is a WIP tool, and is currently alpha software**
 
+TODO:
+
+* module for parsing keyframe files
+* models for each type of deployable artifact
```
Comment (hedss) [inline README.md:8]: For a second I thought you'd misspelt `modules` ;)

File: package-lock.json [hunk 2]
```diff
@@ -0,0 +1,221 @@
+{
```
Comment (hedss) [inline package-lock.json:1]: We're generally not commiting lockfiles atm.

File: src/lib/cli/keyfctl.js [hunk 3]
```diff
@@ -52,133 +52,103 @@ capitano.globalOption({
...
+      keyframe = './keyframe.yml',
+      configuration = './configuration.yml',
+      secrets = './secrets.yml'
+    } = options
+
+    return Promise.join(
+      Keyframe.fromFile(keyframe),
+      Configuration.fromFile(configuration),
+      Secrets.fromFile(secrets),
+      (kf, c, s) => {
```
Comment (hedss) [inline src/lib/cli/keyfctl.js]: A comment, but I'm generally not a fan of varnames less than three characters, if it's not a one/two line use.

File: src/lib/cli/keyfctl.js [hunk 4]
```diff
@@ -52,133 +52,103 @@ capitano.globalOption({
...
+      if (!write) return plan
 
-    if (options.verbose) {
-      res.tap(frames => {
-        console.log('==> All valid frames')
-        for (let frame of frames) {
-          // Output some feedback about the frame history
-          console.log(utils.printFormatFrame(frame))
-        }
+      Promise.map(plan.specs, spec => {
```
Comment (hedss) [inline src/lib/cli/keyfctl.js]: Without a `return`, this map may not have completed by the time the returned action Promise is fulfilled. Just want to ensure that's expected behaviour.

File: src/lib/models/keyframe.js [hunk 5]
```diff
@@ -1,62 +1,193 @@
 'use strict';
 
+const _     = require("lodash")
```
Comment (hedss) [inline src/lib/models/keyframe.js]: Lots of spaces?

**Review by hedss — COMMENTED** [review]
All comments as we're still treating this as a tool in progress (AFAIK?).

File: package-lock.json (hunk 2 above)
Comment (brownjohnf) [reply package-lock.json:1]: Why wouldn't we commit lockfiles?

File: src/lib/models/keyframe.js (hunk 5 above)
Comment (brownjohnf) [reply src/lib/models/keyframe.js]: Ah, leftover from when where were more requires I think.

File: src/lib/cli/keyfctl.js (hunk 4 above)
Comment (brownjohnf) [reply src/lib/cli/keyfctl.js]: I don't *think* it matters here, but better to return.

File: package-lock.json (hunk 2 above)
Comment (hedss) [reply package-lock.json:1]: There were various issues when NPM5 came out, whereby 'expected behaviour' between `package.json` and `package-lock.json` varied massively.

However, I was finding an URL for you, and came across the fact they've fixed this by allowing `package.json` to overrule the lockfile. So, I guess, there isn't the huge objection anymore.

**Review by hedss — APPROVED** [review]

**Pull request merged**